### PR TITLE
Add query params to delegates page

### LINKED ISSUE #8622: WCA Delegates Page: Add query param when switching between regions such that on refresh, you see the same list you were seeing.
**Is your feature request related to a problem? Please describe.**

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by maxidragon**
Closes #8622

### CODE REVIEW

File: WcaOnRails/app/webpacker/components/Delegates/index.jsx
Comment (danieljames-dj) [inline WcaOnRails/app/webpacker/components/Delegates/index.jsx:42]: Why do we have the '+' sign?

File: WcaOnRails/app/webpacker/components/Delegates/index.jsx
Comment (danieljames-dj) [inline WcaOnRails/app/webpacker/components/Delegates/index.jsx:44]: Will this work if `delegateRegions` is empty? Like while the page is loading, `delegateRegions` will be empty right?

**Review by danieljames-dj — COMMENTED** [review]
Thanks @maxidragon for working on this. I'm little busy so couldn't review it completely. I'll review it completely later, or in case Gregor reviews before me, this will definitely be good to go.

File: WcaOnRails/app/webpacker/components/Delegates/index.jsx
Comment (maxidragon) [reply WcaOnRails/app/webpacker/components/Delegates/index.jsx:42]: Because `+` is a short version of `parseInt`. I used region.id as hash and I need to parse it before checking if it equals to `region.id`.

File: WcaOnRails/app/webpacker/components/Delegates/index.jsx
Comment (maxidragon) [reply WcaOnRails/app/webpacker/components/Delegates/index.jsx:44]: If `delegateRegions` is empty, it will show the `Loading` component, I think it is expected behavior

File: WcaOnRails/app/webpacker/components/Delegates/index.jsx
Comment (danieljames-dj) [inline WcaOnRails/app/webpacker/components/Delegates/index.jsx:56]: Instead of find, we can use the selectedRegionIndex from above right?

File: WcaOnRails/app/webpacker/components/Delegates/index.jsx
Comment (danieljames-dj) [inline WcaOnRails/app/webpacker/components/Delegates/index.jsx:48]: selectedSection is an item of delegateRegions and delegateRegions is fetched from server. But delegateRegions doesn’t have component key. Correct me if I’m wrong.

File: WcaOnRails/app/webpacker/components/Delegates/index.jsx
Comment (coder13) [reply WcaOnRails/app/webpacker/components/Delegates/index.jsx:42]: Does this do proper error handling @maxidragon?

File: WcaOnRails/app/webpacker/components/Delegates/index.jsx
Comment (coder13) [inline WcaOnRails/app/webpacker/components/Delegates/index.jsx:118]: What does this do?

File: WcaOnRails/app/webpacker/components/Delegates/index.jsx
Comment (maxidragon) [reply WcaOnRails/app/webpacker/components/Delegates/index.jsx:42]: @coder13 yes, look at lines 43-45
```js
if (selectedRegionIndex === -1) {
      setHash(delegateRegions[0]?.id);
 }
```

File: WcaOnRails/app/webpacker/components/Delegates/index.jsx
Comment (maxidragon) [reply WcaOnRails/app/webpacker/components/Delegates/index.jsx:56]: If we want to use `selectedRegionIndex`, it has to be outside `SelectedComponent`. Did you mean that?

File: WcaOnRails/app/webpacker/components/Delegates/index.jsx
Comment (maxidragon) [reply WcaOnRails/app/webpacker/components/Delegates/index.jsx:118]: It renders a component responsible for setting the default value. I followed what Daniel did in #8529

Comment (danieljames-dj) [reply WcaOnRails/app/webpacker/components/Delegates/index.jsx:113]: The case in that PR is different. There, the component varies depending on the hash. But here, the component to be rendered is always the same - `DelegatesOfRegion`, which is already rendered. So, I guess this line is not needed. If you are also convinced, can you please remove this line and the initialization of `SelectedComponent`?

File: WcaOnRails/app/webpacker/components/Delegates/index.jsx
Comment (danieljames-dj) [reply WcaOnRails/app/webpacker/components/Delegates/index.jsx:51]: Yes, but according to the comment below, you won't be needing `SelectedComponent`. If that's the case, then this comment can be ignored.

File: WcaOnRails/app/webpacker/components/Delegates/index.jsx
Comment (danieljames-dj) [inline WcaOnRails/app/webpacker/components/Delegates/index.jsx:44]: I was about to approve the PR, and suddenly thought of this edge case. Can you please confirm if http://localhost:3000/delegates#europe is opening the Europe page?

Sorry that I couldn't try it myself because my local DB is bit messed up because of which I can't open anything that uses `roles` table as my local DB has `user_roles` instead of `roles`.

Comment (maxidragon) [reply WcaOnRails/app/webpacker/components/Delegates/index.jsx:44]: I used `region.id` as hash, so http://localhost:3000/delegates#3 is opening Europe page. Should I use region name instead of id?

Comment (danieljames-dj) [reply WcaOnRails/app/webpacker/components/Delegates/index.jsx:44]: Oops, just realized that there is no `friendly-id` field in metadata. So you can go ahead with ID itself. Will change once the `friendly_id` is added to metadata.

File: WcaOnRails/app/webpacker/components/Delegates/index.jsx
Comment (maxidragon) [inline WcaOnRails/app/webpacker/components/Delegates/index.jsx:43]: I realized that we need to check if `delegateRegions` is not empty; otherwise, `activeRegion` will always be the first element of the `delegateRegions` array.

**Comment by maxidragon** [comment]
Everything works now ;)
[Screencast from 2023-12-23 14-50-32.webm](https://github.com/thewca/worldcubeassociation.org/assets/94986692/9fa154fc-92f8-4a5f-abd5-ee763de4615c)

**Comment by danieljames-dj** [comment]
> Everything works now ;)
> 
> Video: [Screencast from 2023-12-23 14-50-32.webm](https://github.com/thewca/worldcubeassociation.org/assets/94986692/9fa154fc-92f8-4a5f-abd5-ee763de4615c)

Thanks @maxidragon. All changes looks good to me. Let's wait for a final review from @gregorbg as well.

**Comment by danieljames-dj** [comment]
I've added the friendly ID to the regions in https://github.com/thewca/worldcubeassociation.org/pull/8645. If that gets merged before this PR, you should be able to use `region.metadata.friendly_id` instead of `region.id`. I'll update you in case the other PR gets merged earlier, if not it can be changed later as well.

**Comment by maxidragon** [comment]
#8645 has already been merged, so I changed the `region.id` to `region.metadata.friendly_id `.

**Review by danieljames-dj — APPROVED** [review]
LGTM

**Pull request merged**
